File: Trabalhos Praticos/tp02/TP02Q01 - Classe em Java/Personagem.java
//----------------------------------------------------------------------------------------//
// Avisos
//----------------------------------------------------------------------------------------//
// Olá, esse código provavelmente não está organizado como eu gostaria,
// por conta da urgência das entregas.
// E é normal ter muitas linhas de código comentadas, 
// já que é para uma atividade, eu aproveito isso para documentar todo o estudo também.
//----------------------------------------------------------------------------------------//

import java.util.Scanner;
import java.io.File;
import java.io.RandomAccessFile;
import java.util.Date;
import java.text.SimpleDateFormat;

class Lista {
    String[] codenomes = new String[10];

    public Lista() {        
        // Inicializando o array com strings vazias
        for (int i = 0; i < codenomes.length; i++) {
            codenomes[i] = " ";
        }
    }

    public Lista(String nomes) {
        int tam = nomes.length();
        int size = 0;

        for (int i = 0; i < tam; i++) {
            if (nomes.charAt(i) == ',') {
                size++;
            }
        }
        size++;
        StringBuilder nomesedit = new StringBuilder();
        // String[] finalnomes = new String[size];

        int j = 0;
        // for (int j = 0; j < size; j++) {
        for (int i = 0; i < tam; i++) {
            if (nomes.charAt(i) != '\'' && nomes.charAt(i) != '[' && nomes.charAt(i) != ']') {
                if (nomes.charAt(i) != ',') {
                    nomesedit.append(nomes.charAt(i));
                } else {
                    nomesedit.append(nomes.charAt(i));
                    this.codenomes[j] = nomesedit.toString();
                    nomesedit.setLength(0);
                    j++;
                }
            }
        }
        // }
        this.codenomes[j] = nomesedit.toString();
    }

    // public static void imprime(Lista lista) {
    // // PROBLEMA NO TAMANHO AQUI
    // for (int i = 0; i < 10; i++) {
    // if (lista.codenomes[i] != null) { // Gambiarra
    // System.out.println(lista.codenomes[i]);
    // }
    // }
    // }
}

class Personagem {
    private String id;
    private String name;
    private Lista alternate_name;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private boolean hogwartsStaff;
    private boolean hogwartsStudent;
    private String actorName;
    private boolean alive;
    // alternate_actors
    private Date dateOfBirth;
    private int yearOfBirth;
    private String eyeColour;
    private String gender;
    private String hairColour;
    private boolean wizard;

    public Personagem(String id, String name, Lista alternate_name, String house, String ancestry, String species,
            String patronus, boolean hogwartsStaff, boolean hogwartsStudent, String actorName, boolean alive,
            Date dateOfBirth, int yearOfBirth, String eyeColour, String gender, String hairColour, boolean wizard) {
        this.id = id;
        this.name = name;
        this.alternate_name = alternate_name;
        this.house = house;
        this.ancestry = ancestry;
        this.species = species;
        this.patronus = patronus;
        this.hogwartsStaff = hogwartsStaff;
        this.hogwartsStudent = hogwartsStudent;
        this.actorName = actorName;
        this.alive = alive;
        this.dateOfBirth = dateOfBirth;
        this.yearOfBirth = yearOfBirth;
        this.eyeColour = eyeColour;
        this.gender = gender;
        this.hairColour = hairColour;
        this.wizard = wizard;
    }

    public Personagem() {
        id = " ";
        name = " ";
        alternate_name = new Lista();
        house = " ";
        ancestry = " ";
        species = " ";
        patronus = " ";
        hogwartsStaff = false;
        hogwartsStudent = false;
        actorName = " ";
        alive = false;
        dateOfBirth = null;
        yearOfBirth = -1;
        eyeColour = " ";
        gender = " ";
        hairColour = " ";
        wizard = false;
    }

    // public Personagem() {
    // id = null;
    // name = null;
    // alternate_name = null;
    // house = null;
    // ancestry = null;
    // species = null;
    // patronus = null;
    // hogwartsStaff = false;
    // hogwartsStudent = false;
    // actorName = null;
    // alive = false;
    // dateOfBirth = null;
    // yearOfBirth = -1;
    // eyeColour = null;
    // gender = null;
    // hairColour = null;
    // wizard = false;
    // }

    // ---------- Metodos GET/SET INICIO -----------//
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Lista getAlternate_name() {
        return alternate_name;
    }

    public void setAlternate_name(Lista alternate_name) {
        this.alternate_name = alternate_name;
    }

    public String getHouse() {
        return house;
    }

    public void setHouse(String house) {
        this.house = house;
    }

    public String getAncestry() {
        return ancestry;
    }

    public void setAncestry(String ancestry) {
        this.ancestry = ancestry;
    }

    public String getSpecies() {
        return species;
    }

    public void setSpecies(String species) {
        this.species = species;
    }

    public String getPatronus() {
        return patronus;
    }

    public void setPatronus(String patronus) {
        this.patronus = patronus;
    }

    public boolean isHogwartsStaff() {
        return hogwartsStaff;
    }

    public void setHogwartsStaff(boolean hogwartsStaff) {
        this.hogwartsStaff = hogwartsStaff;
    }

    public boolean isHogwartsStudent() {
        return hogwartsStudent;
    }

    public void setHogwartsStudent(boolean hogwartsStudent) {
        this.hogwartsStudent = hogwartsStudent;
    }

    public String getActorName() {
        return actorName;
    }

    public void setActorName(String actorName) {
        this.actorName = actorName;
    }

    public boolean isAlive() {
        return alive;
    }

    public void setAlive(boolean alive) {
        this.alive = alive;
    }

    public Date getDateOfBirth() {
        return dateOfBirth;
    }

    public void setDateOfBirth(Date dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }

    public int getYearOfBirth() {
        return yearOfBirth;
    }

    public void setYearOfBirth(int yearOfBirth) {
        this.yearOfBirth = yearOfBirth;
    }

    public String getEyeColour() {
        return eyeColour;
    }

    public void setEyeColour(String eyeColour) {
        this.eyeColour = eyeColour;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getHairColour() {
        return hairColour;
    }

    public void setHairColour(String hairColour) {
        this.hairColour = hairColour;
    }

    public boolean isWizard() {
        return wizard;
    }

    public void setWizard(boolean wizard) {
        this.wizard = wizard;
    }

    // ---------- Metodos GET/SET FINAL -----------//
    // ---------- Outros Metodos INICIO -----------//

    public static boolean toUSbool(String frase) {
        if (frase.equals("VERDADEIRO")) {
            return true;
        } else {
            return false;
        }
    }

    public static void imprimirlista(Lista lista) {
        // PROBLEMA NO TAMANHO AQUI
        System.out.print("{");
        for (int i = 0; i < 10; i++) {
            if (lista.codenomes[i] != null) { // Gambiarra?
                System.out.print(lista.codenomes[i]);
            }
        }
        System.out.print("}");
        System.out.print(" ## ");

    }

    public static void imprimir(Personagem p) {
        SimpleDateFormat sdfOutput = new SimpleDateFormat("dd-MM-yyyy");

        System.out.print("[" + p.id + " ## ");
        System.out.print(p.name + " ## ");
        imprimirlista(p.alternate_name);
        System.out.print(p.house + " ## ");
        System.out.print(p.ancestry + " ## ");
        System.out.print(p.species + " ## ");
        System.out.print(p.patronus + " ## ");
        System.out.print(p.hogwartsStaff + " ## ");
        System.out.print(p.hogwartsStudent + " ## ");
        System.out.print(p.actorName + " ## ");
        System.out.print(p.alive + " ## ");
        System.out.print(sdfOutput.format(p.dateOfBirth) + " ## ");
        System.out.print(p.yearOfBirth + " ## ");
        System.out.print(p.eyeColour + " ## ");
        System.out.print(p.gender + " ## ");
        System.out.print(p.hairColour + " ## ");
        System.out.println(p.wizard + "]");

        // System out pra imprimit tudo com uma linha dentro do ()
    }

    public static Personagem clonar(Personagem a) {
        // Personagem b;
        // b.clone(a);
        return a;
    }

    public static void ler(String linhaID) {
        String linha;
        String[] vetor = new String[18];
        SimpleDateFormat sdfInput = new SimpleDateFormat("dd-MM-yyyy");
        SimpleDateFormat sdfOutput = new SimpleDateFormat("dd-MM-yyyy");

        try {
            File file = new File("/tmp/characters.csv");
            Scanner scanner = new Scanner(file);

            // RandomAccessFile file = new RandomAccessFile("/tmp/characters.csv", "rw");
            /// /tmp/characters.csv

            // file.readLine(); // Descarta a primeira linha
            // System.out.println(linha); // testa

            // while ((linha = file.readLine()) != null) {

            scanner.nextLine(); // Descarta a primeira linha
            while ((linha = scanner.nextLine()) != null) {
                vetor = separarlinha(linha);

                String id = vetor[0];

                if (linhaID.equals(id)) {

                    String name = vetor[1];
                    Lista alternate_name = new Lista(vetor[2]);
                    // System.out.println(vetor[2]); // testa
                    // System.out.println(alternate_name.codenomes[3]); // testa
                    String house = vetor[3];
                    String ancestry = vetor[4];
                    String species = vetor[5];
                    String patronus = vetor[6];
                    boolean hogwartsStaff = toUSbool(vetor[7]);
                    boolean hogwartsStudent = toUSbool(vetor[8]);
                    String actorName = vetor[9];
                    boolean alive = toUSbool(vetor[10]);
                    // alternate_actors VETOR 11 NULO
                    Date dateOfBirth = sdfInput.parse(vetor[12]);
                    // String teste = sdfOutput.format(dateOfBirth);
                    int yearOfBirth = Integer.parseInt(vetor[13]);

                    String eyeColour = vetor[14];
                    String gender = vetor[15];
                    String hairColour = vetor[16];

                    boolean wizard = toUSbool(vetor[17]);

                    Personagem p = new Personagem(id, name, alternate_name, house, ancestry,
                            species,
                            patronus, hogwartsStaff, hogwartsStudent, actorName, alive,
                            dateOfBirth, yearOfBirth, eyeColour, gender, hairColour, wizard);

                    imprimir(p);
                    break;
                }
            }
            // file.close();
            scanner.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    // public static Personagem guardar() { // Fiz para estudo
    // String linha;
    // String[] vetor = new String[18];
    // SimpleDateFormat sdfInput = new SimpleDateFormat("dd-MM-yyyy");
    // SimpleDateFormat sdfOutput = new SimpleDateFormat("dd-MM-yyyy");
    // Personagem erro = new Personagem();// ?????????

    // try {
    // RandomAccessFile file = new RandomAccessFile("characters.csv", "rw");
    // file.readLine(); // Descarta a primeira linha
    // // System.out.println(linha); // testa

    // // while (file.getFilePointer() < file.length()) {
    // linha = file.readLine();

    // vetor = separarlinha(linha);
    // // System.out.println(vetor[1]); // testa

    // String id = vetor[0];
    // String name = vetor[1];
    // Lista alternate_name = new Lista(vetor[2]);
    // // System.out.println(vetor[2]); // testa
    // // System.out.println(alternate_name.codenomes[3]); // testa
    // String house = vetor[3];
    // String ancestry = vetor[4];
    // String species = vetor[5];
    // String patronus = vetor[6];
    // boolean hogwartsStaff = toUSbool(vetor[7]);
    // boolean hogwartsStudent = toUSbool(vetor[8]);
    // String actorName = vetor[9];
    // boolean alive = toUSbool(vetor[10]);
    // // alternate_actors VETOR 11 NULO
    // Date dateOfBirth = sdfInput.parse(vetor[12]);
    // // String teste = sdfOutput.format(dateOfBirth);
    // // System.out.println(teste); // testa
    // int yearOfBirth = Integer.parseInt(vetor[13]);

    // String eyeColour = vetor[14];
    // String gender = vetor[15];
    // String hairColour = vetor[16];

    // boolean wizard = toUSbool(vetor[17]);

    // Personagem p = new Personagem(id, name, alternate_name, house, ancestry,
    // species,
    // patronus, hogwartsStaff, hogwartsStudent, actorName, alive,
    // dateOfBirth, yearOfBirth, eyeColour, gender, hairColour, wizard);

    // file.close();
    // return p;
    // // }
    // } catch (Exception e) {
    // e.printStackTrace();
    // }
    // return erro; // ???????
    // }

    public static String[] separarlinha(String linha) {
        StringBuilder res = new StringBuilder();
        String[] resfinal = new String[18];
        int tam = linha.length();

        int j = 0;
        for (int i = 0; i < tam; i++) {
            if (linha.charAt(i) != ';') {
                res.append(linha.charAt(i));
            } else {
                resfinal[j] = res.toString();
                res.setLength(0);
                j++;
            }
        }
        resfinal[j] = res.toString();
        // System.out.println(resfinal[2]); // testa
        return resfinal;
    }

    // ---------- Outros Metodos FINAL -----------//
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // while (sc.hasNextLine()) {
        String ID = sc.nextLine();

        while (ID.equals("FIM") != true) {
            ler(ID);
            ID = sc.nextLine();
        }
        sc.close();
    }
}

File: Trabalhos Praticos/tp02/TP02Q14 - Radixsort em C/Personagem.c
//----------------------------------------------------------------------------------------//
// Avisos
//----------------------------------------------------------------------------------------//
// Olá, esse código provavelmente não está organizado como eu gostaria,
// por conta da urgência das entregas.
// E é normal ter muitas linhas de código comentadas,
// já que é para uma atividade, eu aproveito isso para documentar todo o estudo também.
//----------------------------------------------------------------------------------------//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef struct Lista
{
    char codenomes[10][256]; // Supondo que cada nome não exceda 255 caracteres
} Lista;

void inicializarLista(Lista *lista)
{
    for (int i = 0; i < 10; i++)
    {
        strcpy(lista->codenomes[i], "VAZIO"); // Inicializa todos os nomes como strings vazias
    }
}

typedef struct Personagem
{
    char id[256];
    char name[256];
    Lista alternate_name;
    char house[256];
    char ancestry[256];
    char species[256];
    char patronus[256];
    int hogwartsStaff;   // bool
    int hogwartsStudent; // bool
    char actorName[256];
    int alive;            // bool
    char dateOfBirth[11]; // Formato "dd-mm-yyyy"
    int yearOfBirth;
    char eyeColour[256];
    char gender[256];
    char hairColour[256];
    int wizard; // bool
} Personagem;

int toUSbool(const char *frase)
{
    // return strcmp(frase, "VERDADEIRO") == 0;

    // const char *resultado = strstr(frase, "VERDADEIRO");
    // return resultado != NULL;
    return strstr(frase, "VERDADEIRO") != NULL;
}

void imprimirLista(const Lista *lista)
{
    printf("{");
    int i;
    for (i = 0; i < 10; i++)
    {
        if (strcmp(lista->codenomes[i], "VAZIO") != 0)
        {
            printf("%s", lista->codenomes[i]);
            // Verifica se não é o último elemento e se o próximo não está vazio
            if (i < 8 && strcmp(lista->codenomes[i + 1], "VAZIO") != 0)
            {
                printf(","); // Adiciona uma vírgula apenas se houver próximo elemento
            }
        }
    }
    printf("} ## ");
}

void imprimir(const Personagem *p)
{
    printf("[%s ## %s ## ", p->id, p->name);
    imprimirLista(&p->alternate_name);
    printf("%s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %d ## %s ## %s ## %s ## %s]\n",
           p->house, p->ancestry, p->species, p->patronus,
           p->hogwartsStaff ? "true" : "false",   // Diretamente usando o valor booleano
           p->hogwartsStudent ? "true" : "false", // Diretamente usando o valor booleano
           p->actorName,
           p->alive ? "true" : "false", // Diretamente usando o valor booleano
           p->dateOfBirth,
           p->yearOfBirth,
           p->eyeColour, p->gender, p->hairColour,
           p->wizard ? "true" : "false"); // Diretamente usando o valor booleano
}

// Função para remover caracteres especiais de uma string
void limparCaracteresEspeciais(char *string)
{
    int i, j;
    for (i = 0, j = 0; string[i] != '\0'; i++)
    {
        // Verifica se o caractere atual é especial
        if (string[i] != '\'' && string[i] != '[' && string[i] != ']')
        {
            // Se não for especial, copia para a nova posição
            string[j++] = string[i];
        }
    }
    // Adiciona o terminador nulo à nova posição final
    string[j] = '\0';
}

// Função para ler e processar os nomes alternativos
void processarNomesAlternativos(char *nomesAlternativos, Lista *lista)
{
    inicializarLista(lista);
    char *nome = strtok(nomesAlternativos, ",");
    int i = 0;
    while (nome != NULL && i < 10)
    {
        // Remove caracteres especiais do nome alternativo
        limparCaracteresEspeciais(nome);
        // Copia o nome limpo para a lista de codenomes
        strcpy(lista->codenomes[i++], nome);
        // Proximo token
        nome = strtok(NULL, ",");
    }
}

void ler(Personagem *p, char *ID)
{

    FILE *file = fopen("/tmp/characters.csv", "r");
    // FILE *file = fopen("./characters.csv", "r");
    if (!file)
    {
        perror("Erro ao abrir o arquivo");
        return;
    }

    char linha[1024];

    while (fgets(linha, sizeof(linha), file))
    {
        char vetor[18][256];
        int index = 0;
        char *token = linha;
        char *next_token;

        while (token)
        {
            next_token = strchr(token, ';');
            if (next_token)
            {
                *next_token = '\0'; // Termina o token atual
                strcpy(vetor[index++], token);
                token = next_token + 1;
            }
            else
            {
                strcpy(vetor[index++], token);
                break; // Não há mais delimitadores, finaliza o loop
            }
        }

        if (strcmp(ID, vetor[0]) == 0)
        {
            strcpy(p->id, vetor[0]);
            strcpy(p->name, vetor[1]);

            // Trata a Lista
            processarNomesAlternativos(vetor[2], &p->alternate_name);

            // Copia as outras informações
            strcpy(p->house, vetor[3]);
            strcpy(p->ancestry, vetor[4]);
            strcpy(p->species, vetor[5]);
            strcpy(p->patronus, vetor[6]);
            p->hogwartsStaff = toUSbool(vetor[7]);
            p->hogwartsStudent = toUSbool(vetor[8]);
            strcpy(p->actorName, vetor[9]);
            p->alive = toUSbool(vetor[10]);
            strcpy(p->dateOfBirth, vetor[12]); // Supondo que a data esteja formatada corretamente
            p->yearOfBirth = atoi(vetor[13]);
            strcpy(p->eyeColour, vetor[14]);
            strcpy(p->gender, vetor[15]);
            strcpy(p->hairColour, vetor[16]);
            p->wizard = toUSbool(vetor[17]);

            break;
        }
    }
    fclose(file);
}

int getDigit(const char *id, int pos)
{
    return id[pos];
}

void radixSort(Personagem **vetor, int n, int tam, int info[]) {
    Personagem **temp = (Personagem **)malloc(n * sizeof(Personagem *));
    if (temp == NULL) {
        printf("Erro ao alocar memória.\n");
        return;
    }

    for (int pos = tam - 1; pos >= 0; pos--) {
        int count[256] = {0};

        for (int i = 0; i < n; i++) {
            int digit = getDigit(vetor[i]->id, pos);
            count[digit]++;
            info[0]++;
        }

        for (int i = 1; i < 256; i++) {
            count[i] += count[i - 1];
        }

        for (int i = n - 1; i >= 0; i--) {
            int digit = getDigit(vetor[i]->id, pos);
            temp[--count[digit]] = vetor[i];
            info[1]++; 
        }

        for (int i = 0; i < n; i++) {
            vetor[i] = temp[i];
        }
    }
    free(temp);
}

void registraLog(int tempoExecucao, int info[])
{
    FILE *logFile;
    double segundos = (double)tempoExecucao / 1000000000.0; // Convertendo para segundos

    logFile = fopen("831566_radixSort.txt", "a");

    if (logFile == NULL)
    {
        printf("Erro ao abrir o arquivo.\n");
        return;
    }

    fprintf(logFile, "831566\t%d\t%d\t%f\n", info[0], info[1], segundos);

    fclose(logFile);
}

int main()
{
    clock_t inicio, fim;
    double tempoExecucao;
    inicio = clock();
    int info[2] = {0}; // info[0] = comparacoes, info[1] = movimentacoes

    Personagem *personagens[30];

    char ID[256];
    int count = 0;

    // scanf(" %[^\n]", ID);
    scanf(" %[^\r\n]s", ID);

    while (strcmp(ID, "FIM") != 0)
    {
        // ID[strcspn(ID, "\n\r")] = 0;

        ID[strcspn(ID, "\n")] = 0; // Remove newline
        personagens[count] = (Personagem *)malloc(sizeof(Personagem));
        ler(personagens[count], ID);

        count++;
        scanf(" %[^\r\n]s", ID);
    }

    radixSort(personagens, count, 256, info);

    for (int i = 0; i < count; i++)
    {
        imprimir(personagens[i]);
        free(personagens[i]);
    }

    fim = clock();
    tempoExecucao = (double)(fim - inicio);
    registraLog(tempoExecucao, info);
    return 0;
}